Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -7,326 +7,390 @@
-			return this._PendingTip;
-		}
-
-		public ChainedBlock PendingTip
-		{
-			get
-			{
-				var tip = this._PendingTip;
-				if (tip == null)
-					return null;
-				//Prevent memory leak by returning a block from the chain instead of real pending tip of possible
-				return this.Chain.GetBlock(tip.HashBlock) ?? tip;
-			}
-		}
-
-		protected override void DetachCore()
-		{
-            this.AttachedNode.StateChanged -= this.AttachedNode_StateChanged;
-		}
+            this.RegisterDisposable(this.AttachedNode.Filters.Add(this.Intercept));
+
+            this.logger.LogTrace("(-)");
+        }
 
+        protected override void DetachCore()
+        {
+            this.logger.LogTrace("()");
 
-	    #region ICloneable Members
+            this.AttachedNode.StateChanged -= this.AttachedNode_StateChanged;
 
-		public override object Clone()
-		{
-			var clone = new ChainHeadersBehavior(this.Chain, this.SharedState)
-			{
-				CanSync = this.CanSync,
-				CanRespondToGetHeaders = this.CanRespondToGetHeaders,
-				AutoSync = this.AutoSync,
-				_State = this._State
-			};
-			return clone;
-		}
+            this.logger.LogTrace("(-)");
+        }
+
+        private void Intercept(IncomingMessage message, Action act)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(message), message.Message.Command, nameof(this.AttachedNode), this.AttachedNode?.RemoteSocketEndpoint);
+
+            var inv = message.Message.Payload as InvPayload;
+            if (inv != null)
+            {
+                if (inv.Inventory.Any(i => ((i.Type & InventoryType.MSG_BLOCK) != 0) && !this.Chain.Contains(i.Hash)))
+                {
+                    // No need of periodical refresh, the peer is notifying us.
+                    this.refreshTimer.Dispose(); 
+                    if (this.AutoSync)
+                        this.TrySync();
+                }
+            }
+
+            // == GetHeadersPayload ==
+            // Represents our height from the peer's point of view. 
+            // It is sent from the peer on first connect, in response to Inv(Block) 
+            // or in response to HeaderPayload until an empty array is returned.
+            // This payload notifies peers of our current best validated height. 
+            // Use the ChainState.ConsensusTip property (not Chain.Tip)
+            // if the peer is behind/equal to our best height an empty array is sent back.
+
+            // Ignoring "getheaders" from peers because node is in initial block download.
+            var getheaders = message.Message.Payload as GetHeadersPayload;
+            if ((getheaders != null) 
+                && this.CanRespondToGetHeaders
+                // If not in IBD whitelisted won't be checked.
+                && (!this.chainState.IsInitialBlockDownload || this.AttachedNode.Behavior<ConnectionManagerBehavior>().Whitelisted)) 
```

Comment:
A suggested improvement, I think this will never change so we could set a property with this behaviour 
`this.AttachedNode.Behavior<ConnectionManagerBehavior>().Whitelisted))`

Reply:
possibly true, but offtopic ;)